File: src/main/scala/com/lemur/utils/GoogleVerifier.java
package com.lemur.utils;

import com.google.api.client.googleapis.auth.oauth2.GoogleIdToken;
import com.google.api.client.googleapis.auth.oauth2.GoogleIdTokenVerifier;
import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
import com.google.api.client.http.javanet.NetHttpTransport;
import com.google.api.client.json.jackson2.JacksonFactory;

import java.io.IOException;
import java.security.GeneralSecurityException;
import java.util.Arrays;

/**
 * Created by djhurley on 02/04/16.
 */
public class GoogleVerifier {
    public static boolean verify(String idTokenString) {
        try {
            NetHttpTransport transport = GoogleNetHttpTransport.newTrustedTransport();
            JacksonFactory jsonFactory = JacksonFactory.getDefaultInstance();
            String clientId = "822785640622-f0tj5dvauvn97hieujjpbtfbffvr5uvc.apps.googleusercontent.com";
            GoogleIdTokenVerifier verifier = new GoogleIdTokenVerifier.Builder(transport, jsonFactory)
                    .setAudience(Arrays.asList(clientId))
                    .setIssuer("https://accounts.google.com")
                    .build();

            GoogleIdToken idToken = verifier.verify(idTokenString);

            if (idToken != null) {
                GoogleIdToken.Payload payload = idToken.getPayload();

                System.out.println(payload);

                return true;

            } else {
                System.out.println("Invalid ID token: " + idTokenString);
            }
        }catch(IOException e){
            System.out.println(e.getMessage());
        }catch(GeneralSecurityException e){
            System.out.println(e.getMessage());
        }catch(IllegalArgumentException e) {
            System.out.println(e.getMessage());
        }

        return false;
    }
}

File: src/main/scala/com/lemur/utils/GoogleVerifier.scala
package com.lemur.utils

import com.google.api.client.googleapis.auth.oauth2.GoogleIdTokenVerifier
import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport
import com.google.api.client.json.jackson2.JacksonFactory
import com.lemur.common.Constants.googleIssuer
import com.typesafe.config.ConfigFactory

import scala.collection.JavaConversions._
import scala.collection.mutable.ArrayBuffer

trait Verifier {

  def verify(idTokenString: String): Boolean
}

object GoogleVerifier extends Verifier {

  override def verify(idTokenString: String): Boolean = {

    val clientId = ConfigFactory.load().getString("authentication.client.id")
    val transport = GoogleNetHttpTransport.newTrustedTransport
    val jsonFactory = JacksonFactory.getDefaultInstance
    val verifier = new GoogleIdTokenVerifier.Builder(transport, jsonFactory)
      .setIssuer(googleIssuer).setAudience(ArrayBuffer(clientId)).build()
    val idToken = Option(verifier.verify(idTokenString))

    idToken match {
      case Some(token) =>
        println(token.getPayload)
        true

      case None =>
        println("Invalid ID token: " + idTokenString)
        false
    }
  }
}
